lib/tasks/nokul/archive_prospective_students.rake

```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+namespace :nokul do
+  desc 'Archive prospective students if expiry date has passed'
+  task archive_prospective_students: :environment do
+    ProspectiveStudent.not_archived.find_each { |ps| ps.update(archived: true) if ps.expiry_date < Time.zone.now }
```

Comment:
Burada ciddi bir performans sorunu var, tipik bir N+1 durumu. Tek bir kaydı güncellemek için çalışan SQL sorgusuna bakalım:

```
[1 -> SELECT] ProspectiveStudent Load (0.9ms)  SELECT "prospective_students".* FROM "prospective_students" WHERE "prospective_students"."archived" = $1 ORDER BY "prospective_students"."id" ASC LIMIT $2  [["archived", false], ["LIMIT", 1000]]
[2 -> SELECT] AcademicTerm Load (0.2ms)  SELECT "academic_terms".* FROM "academic_terms" WHERE "academic_terms"."id" = $1 LIMIT $2  [["id", 3], ["LIMIT", 1]]
[3 -> SELECT] StudentEntranceType Load (0.1ms)  SELECT "student_entrance_types".* FROM "student_entrance_types" WHERE "student_entrance_types"."id" = $1 LIMIT $2  [["id", 40], ["LIMIT", 1]]
[4 -> SELECT] Unit Load (0.1ms)  SELECT "units".* FROM "units" WHERE "units"."id" = $1 LIMIT $2  [["id", 682], ["LIMIT", 1]]
[5 -> SELECT] ProspectiveStudent Exists (0.2ms)  SELECT 1 AS one FROM "prospective_students" WHERE "prospective_students"."id_number" = $1 AND "prospective_students"."id" != $2 AND "prospective_students"."unit_id" = $3 AND "prospective_students"."exam_score" = $4 LIMIT $5  [["id_number", "1"], ["id", 946], ["unit_id", 682], ["exam_score", 437.4], ["LIMIT", 1]]
[6 -> UPDATE] ProspectiveStudent Update (0.5ms)  UPDATE "prospective_students" SET "archived" = $1, "updated_at" = $2 WHERE "prospective_students"."id" = $3  [["archived", true], ["updated_at", "2019-03-14 13:51:06.198015"], ["id", 946]]
```

8500 prospective_student için bu çalıştığında `8500 * 6= 51.000` tane SQL sorgusu çalışacak.

Bu 51.000 sorguyu 1 sorguya düşürebiliriz:

```
    prospective_students = ProspectiveStudent.arel_table
    ProspectiveStudent.not_archived.where(prospective_students[:expiry_date].lt(Time.zone.now)).update_all(archived: true)
```